### PR TITLE
Add a HOC for the REST API tooling

### PR DESCRIPTION
**Pull request by grampelberg**
We're copying and duplicating logic all over the place with components that need to talk to the API.

Moves most of the REST API tooling into a HOC that can be used by other components. Now, a component can use `withREST`, pass in the promises that it would like resolved and receive the responses as props.

### CODE REVIEW

**Comment by grampelberg** [comment]
@rmars @klingerf I'm inclined to keep this to a single component for now and roll it out across the others in a separate PR. What do you think? I'm happy to update this and get it working for everything as well.

File: web/app/js/components/ResourceList.jsx [hunk 1]
```diff
@@ -6,114 +6,48 @@ import MetricsTable from './MetricsTable.jsx';
...
-      numResources={_.size(this.state.metrics)} />);
+      numResources={0} />);
   }
 
   render() {
-    let friendlyTitle = _.startCase(this.props.resource);
+    const {api, controllerNamespace, data, error, loading, resource} = this.props;
+
+    if (error) return  <ErrorBanner message={error} />;
+    if (loading) return <ConduitSpinner />;
```
Comment (klingerf) [inline web/app/js/components/ResourceList.jsx]: Totally trivial but 👍 to @rmars comment in #949 -- I'm inclined to write this as:

```
    if (error) { return <ErrorBanner message={error} />; }
    if (loading) { return <ConduitSpinner />; }
```

File: web/app/js/components/ResourceList.jsx [hunk 2]
```diff
@@ -6,114 +6,48 @@ import MetricsTable from './MetricsTable.jsx';
...
-
-    this.setState({
-      pendingRequests: false,
-      error: `Error getting data from server: ${e.message}`
-    });
-  }
-
-  renderEmptyMessage() {
+export class ResourceList extends React.Component {
+  renderEmptyMessage = () => {
```
Comment (klingerf) [inline web/app/js/components/ResourceList.jsx]: I can never remember the difference, but I think this has something to do with binding the caller? What's the difference between the way this method was previously defined:

```
  renderEmptyMessage() {
    ...
  }
```

And the ES6 version you're using here:

```
  renderEmptyMessage = () => {
    ...
  }
```

Am just curious, and hopefully one day I'll internalize what's going on here 😉

File: web/app/js/components/ResourceList.jsx [hunk 3]
```diff
@@ -6,114 +6,48 @@ import MetricsTable from './MetricsTable.jsx';
...
+    if (error) return  <ErrorBanner message={error} />;
+    if (loading) return <ConduitSpinner />;
+
+    const processedMetrics = processSingleResourceRollup(
+      data[0], controllerNamespace);
+
+    const friendlyTitle = _.startCase(resource);
     return (
       <div className="page-content">
-        { !this.state.error ? null : <ErrorBanner message={this.state.error} /> }
```
Comment (klingerf) [inline web/app/js/components/ResourceList.jsx]: The previous implementation of this component was setup so that, if an API error was encountered, we'd still display the page header, below the error banner. In this branch you're changing it so that we only display the error banner and omit the page header, by returning early on line 25. It now looks like this:

image

I have a slight preference for the previous implementation, so I'd vote to remove the early return and restore this line instead.

Comment (grampelberg) [reply web/app/js/components/ResourceList.jsx]: You're totally right, this is why I hate that syntax style ... it is super hard to parse. Let me give it another go and see what you think.

Comment (grampelberg) [reply web/app/js/components/ResourceList.jsx]: @klingerf I updated and added some tests for `ResourceList` to make sure that is the validated behavior. Mind taking a quick look and making sure that's good?

File: web/app/test/ResourceListTest.jsx [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+import Adapter from 'enzyme-adapter-react-16';
+import CallToAction from '../js/components/CallToAction.jsx';
+import ConduitSpinner from '../js/components/ConduitSpinner.jsx';
+import deployRollup from './fixtures/deployRollup.json';
+import Enzyme from 'enzyme';
+import ErrorBanner from '../js/components/ErrorBanner.jsx';
+import { expect } from 'chai';
+import MetricsTable from '../js/components/MetricsTable.jsx';
+import PageHeader from '../js/components/PageHeader.jsx';
+import { processSingleResourceRollup } from '../js/components/util/MetricUtils.js';
```
Comment (klingerf) [inline web/app/test/ResourceListTest.jsx]: Unused import?

File: web/app/test/ResourceListTest.jsx [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  it('shows a loading spinner', () => {
+    const component = shallow(
+      <ResourceList
+        data={[]}
+        loading={true} />
+    );
+
+    expect(component.find(ErrorBanner)).to.have.length(0);
+    expect(component.find(PageHeader)).to.have.length(1);
+    expect(component.find(ConduitSpinner)).to.have.length(1);
```
Comment (klingerf) [inline web/app/test/ResourceListTest.jsx]: This is minor, but I don't think this test would pass against the current conduit master. Specifically, in the existing implementation, if the `ConduitSpinner` component is present, the `PageHeader` component is not. So to test the existing behavior, these expectations should be:

```
    expect(component.find(ErrorBanner)).to.have.length(0);
    expect(component.find(PageHeader)).to.have.length(0);
    expect(component.find(ConduitSpinner)).to.have.length(1);
    expect(component.find(CallToAction)).to.have.length(0);
    expect(component.find(MetricsTable)).to.have.length(0);
```

I'm ok with changing the behavior to display both the page header and the spinner simultaneously, but I want to make sure we're intentional about it. For instance, if we do go this route, then we should update the `Namespace` and `ServiceMesh` components as well, to make sure we're consistently displaying spinner + page header throughout.

File: web/app/test/ResourceListTest.jsx (hunk 4 above)
Comment (grampelberg) [reply web/app/test/ResourceListTest.jsx]: Good catch. Do we have this in eslint? I feel like some kind of tool should have yelled at me about it.

File: web/app/test/ResourceListTest.jsx (hunk 5 above)
Comment (grampelberg) [reply web/app/test/ResourceListTest.jsx]: Nope, you're totally right. This is my bad on reading the code again. Once the jest stuff lands, I want to spend some time getting the tests a little bit better. I've realized I am relying on them to yell at me a little too much given their current state.

File: web/app/test/ResourceListTest.jsx (hunk 4 above)
Comment (klingerf) [reply web/app/test/ResourceListTest.jsx]: Good question. Looks like our current eslint config catches this, but it's set to warn instead of error. So I think you would have seen the warning when running the tests, but the tests would still pass.

```
    "no-unused-vars": [1, { "args": "all", "argsIgnorePattern": "^_" }],
```

I'm working on an eslint branch, and I'll probably move all of our warn rules to error, to make it more obvious when code is in violation of one of the rules.
